upload-api/did-resolver.js

```diff
@@ -0,0 +1,58 @@
+import { Schema, DIDResolutionError } from '@ucanto/validator'
+import { ok, error } from '@ucanto/core'
+import { parseDidPlc } from '@storacha/did-plc'
+import * as ed25519 from '@ucanto/principal/ed25519'
+
+/**
+ * @typedef {Record<`did:web:${string}`, `did:key:${string}`>} PrincipalMapping
+ */
+
+/**
+ * @typedef {import('@ucanto/interface').Signature} Signature
+ */
+
+/**
+ * Creates a DID resolver that can resolve web and PLC DIDs to their corresponding DID keys.
+ *
+ * @param {PrincipalMapping} principalMapping - Mapping of known WebDIDs to their corresponding DIDKeys
+ * @param {import('@storacha/did-plc').PlcClient} plcClient - PLC client to use for resolving PLC DIDs
+ * @returns {(did: `did:${string}:${string}`) => Promise<import('@ucanto/interface').Result<`did:key:${string}`[], DIDResolutionError>>}
+ */
+export function createDidResolver(principalMapping, plcClient) {
+  /**
+   * @param {`did:${string}:${string}`} did
+   */
+  return async (did) => {
+    if (Schema.did({ method: 'web' }).is(did) && principalMapping[did]) {
+      return ok([/** @type {`did:key:${string}`} */(principalMapping[did])])
+    }
+
+    if (Schema.did({ method: 'plc' }).is(did) && plcClient) {
+      return resolveDIDPlc(did, plcClient)
+    }
+
+    return error(new DIDResolutionError(did))
+  }
+}
+
+/**
+ * Fetches the DID PLC Document and verifies the delegation signature against 
+ * the verification methods available in the document.
+ * 
+ * @param {`did:${string}:${string}`} did
+ * @param {import('@storacha/did-plc').PlcClient} plcClient
+ * @returns {Promise<import('@ucanto/interface').Result<`did:key:${string}`[], DIDResolutionError>>}
+ */
+const resolveDIDPlc = async (did, plcClient) => {
+  try {
+    const didPlc = parseDidPlc(did)
+    /** @type {import('@storacha/did-plc/types').PlcDocument} */
+    const doc = await plcClient.getDocument(didPlc)
+    const verificationMethods = doc.verificationMethod || []
+    const verifiers = verificationMethods.map(vm => ed25519.Verifier.parse(vm.publicKeyMultibase))
```

Comment:
Hmm, should we check the type and ignore ones we don't recognise (log a warn) rather than erroring altogether?